core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -811,8 +811,36 @@ private[spark] class TaskSchedulerImpl(
     blacklistTrackerOpt.map(_.nodeBlacklist()).getOrElse(scala.collection.immutable.Set())
   }
 
-  // By default, rack is unknown
-  def getRackForHost(value: String): Option[String] = None
+  // Add a on-off switch to save time for rack resolving
+  lazy val skipRackResolving: Boolean = sc.conf.get(LOCALITY_WAIT_RACK) == 0L
+
+  /**
+   * Rack is unknown by default.
+   * It can be override in different TaskScheduler, like Yarn.
+   */
+  def defaultRackValue: Option[String] = None
+
+  def doGetRacksForHosts(preferredLocation: List[String]): List[Option[String]] = Nil
+
+  def getRackForHost(value: String): Option[String] = {
+    if (skipRackResolving) {
+      defaultRackValue
+    } else {
+      doGetRacksForHosts(List(value)) match {
+        case Nil => None
+        case h :: Nil => h
+        case _ => None
+      }
+    }
+  }
+
+  def getRacksForHosts(values: List[String]): List[Option[String]] = {
```

Comment:
I don't think you should use `List` here, for either the argument or return type -- if you're looking for the equivalent of `java.util.List`, its probably Scala's `Seq`.  I think `List` is unnecessarily specific (and probably non-optimal anyway)

Reply:
updated